superset/utils/migrate_viz.py

```diff
@@ -0,0 +1,125 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import json
+from enum import Enum
+from typing import Dict, Set, Type, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from superset.models.slice import Slice
+
+
+# pylint: disable=invalid-name
+class MigrateVizEnum(str, Enum):
+    # the Enum member name is viz_type in database
+    treemap = "treemap"
+
+
+class MigrateViz:
+    remove_keys: Set[str] = set()
+
+    mapping_keys: Dict[str, str] = {}
+
+    source_viz_type: str
+
+    target_viz_type: str
+
+    def __init__(self, form_data: str) -> None:
+        self.data = json.loads(form_data)
+
+    def _pre_action(self) -> None:
+        """some actions before migrate"""
+
+    def _migrate(self) -> None:
+        if self.data.get("viz_type") != self.source_viz_type:
+            return
+
+        if "viz_type" in self.data:
+            self.data["viz_type"] = self.target_viz_type
+
+        rv_data = {}
+        for (key, value) in self.data.items():
+            if key in self.mapping_keys and self.mapping_keys[key] in rv_data:
+                raise ValueError("Duplicate key in target viz")
+
+            if key in self.mapping_keys:
+                rv_data[self.mapping_keys[key]] = value
```

Comment:
I personally find this more readable, but it's not a blocker.

```suggestion
            if key in self.mapping_keys:
                mapped_key = self.mapping_keys[key] 
                if mapped_key in rv_data:
                   raise ValueError("Duplicate key in target viz")
                rv_data[mapped_key] = value          
```

Reply:
Exception handling is easier to maintain if it can be put together.